### ISSUE TITLE
golangci-lint cache does not detect changes in ruleguard rules

### ISSUE THREAD

**Issue by dottedmag** [golangci/golangci-lint#1999]
- [x] Yes, I'm using a binary release within 2 latest major releases. Only such installations are supported.
- [x] Yes, I've searched similar issues on GitHub and didn't find any.
- [x] Yes, I've included all information below (version, config, etc).
- [x] Yes, I've tried with the standalone linter if available. (https://golangci-lint.run/usage/linters/)

<details><summary>golangci-lint cache does not detect changes in ruleguard rules</summary>

* Clone https://github.com/ridge/golangci-cache-ruleguard
* `golangci-lint run`
* Note `ruleguard: A found` linter diagnostic
* Change `rules/rules.go`: comment out `A found`, uncomment `B found` lines.
* `golangci-lint run`
* Note stale `ruleguard: A found` linter diagnostic
* `golangci-lint cache clean && golangci-lint run`
* Note fresh `ruleguard: B found` linter diagnostic

So the cache is not being discarded when ruleguard rules change.

This is problematic:
- In CI it means caching must be disabled, so CI builds become slower than they should be
- For local environment it means a local lint run will return stale results, especially after pulling in changed rules from the repository. This is extremely confusing.

</details>

<details><summary>Version of golangci-lint</summary>

```console
$ golangci-lint --version
golangci-lint has version 1.39.0 built from 9aea4aee on 2021-03-26T08:02:53Z
```

</details>

<details><summary>Config file</summary>

```console
$ cat .golangci.yml
linters:
  enable:
    - gocritic
linters-settings:
  gocritic:
    enabled-checks:
      - ruleguard
    settings:
      ruleguard:
        rules: "rules/rules.go"
```

</details>

<details><summary>Go environment</summary>

```console
$ go version && go env
go version go1.16.3 linux/amd64
GO111MODULE=""
GOARCH="amd64"
GOBIN=""
GOCACHE="/home/dottedmag/.cache/go-build"
GOENV="/home/dottedmag/.config/go/env"
GOEXE=""
GOFLAGS=""
GOHOSTARCH="amd64"
GOHOSTOS="linux"
GOINSECURE=""
GOMODCACHE="/home/dottedmag/go/pkg/mod"
GONOPROXY=""
GONOSUMDB=""
GOOS="linux"
GOPATH="/home/dottedmag/go"
GOPRIVATE=""
GOPROXY="https://proxy.golang.org,direct"
GOROOT="/home/dottedmag/tectonic/_deps.linux-amd64/go-1.16.3"
GOSUMDB="sum.golang.org"
GOTMPDIR=""
GOTOOLDIR="/home/dottedmag/tectonic/_deps.linux-amd64/go-1.16.3/pkg/tool/linux_amd64"
GOVCS=""
GOVERSION="go1.16.3"
GCCGO="gccgo"
AR="ar"
CC="gcc"
CXX="g++"
CGO_ENABLED="1"
GOMOD="/home/dottedmag/w/golangci-cache-ruleguard/go.mod"
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -m64 -pthread -fmessage-length=0 -fdebug-prefix-map=/tmp/go-build2596763710=/tmp/go-build -gno-record-gcc-switches"
```

</details>

<details><summary>Verbose output of running</summary>

```console
$ golangci-lint cache clean
$ golangci-lint run -v
# as explained above
```

</details>

<details><summary>Code example or link to a public repository</summary>

https://github.com/ridge/golangci-cache-ruleguard
</details>

**Comment by SVilgelm** [comment]
Hi, could you please put a correct link? the https://github.com/ridge/golangci-cache-ruleguard repo is not available, 404 error

**Comment by dottedmag** [comment]
Apologies, it was accidentally private. Should work now.

**Comment by thockin** [comment]
I am seeing this still in 1.43.0


Rules:
```
$ cat ruleguard_rules.go 
//+build never

package gorules

import "github.com/quasilyte/go-ruleguard/dsl"

func netParseIP(m dsl.Matcher) {
	m.Match(`net.ParseIP($_)`).Report("prefer utilnet.ParseIPSloppy()")
}
```

Lint config:
```
$ cat .golangci.yaml
run:
  timeout: 30m

linters:
  disable-all: true
  enable:
    - gocritic
linters-settings:
  gocritic:
    enabled-checks:
      - ruleguard
    settings:
      ruleguard:
        rules: "${configDir}/ruleguard_rules.go"
```

Run, works as expected:

```
$ ./_output/local/bin/golangci-lint run -v ./pkg/proxy/...
INFO [config_reader] Config search paths: [./ /home/thockin/src/go/src/k8s.io/kubernetes/pkg/proxy /home/thockin/src/go/src/k8s.io/kubernetes/pkg /home/thockin/src/go/src/k8s.io/kubernetes /home/thockin/src/go/src/k8s.io /home/thockin/src/go/src /home/thockin/src/go /home/thockin/src /home/thockin /home /] 
INFO [config_reader] Used config file .golangci.yaml 
INFO [lintersdb] Active 1 linters: [gocritic]     
INFO [loader] Go packages loading at mode 575 (exports_file|types_sizes|compiled_files|deps|files|imports|name) took 534.192393ms 
INFO [runner/filename_unadjuster] Pre-built 0 adjustments in 19.76924ms 
INFO [linters context/goanalysis] analyzers took 0s with no stages 
INFO [runner] Processors filtering stat (out/in): diff: 1/1, max_same_issues: 1/1, max_from_linter: 1/1, source_code: 1/1, path_shortener: 1/1, autogenerated_exclude: 1/1, nolint: 1/1, uniq_by_line: 1/1, sort_results: 1/1, cgo: 1/1, path_prettifier: 1/1, exclude-rules: 1/1, path_prefixer: 1/1, filename_unadjuster: 1/1, skip_dirs: 1/1, identifier_marker: 1/1, severity-rules: 1/1, skip_files: 1/1, exclude: 1/1, max_per_file_from_linter: 1/1 
INFO [runner] processing took 3.204879ms with stages: nolint: 2.93757ms, source_code: 92.342µs, autogenerated_exclude: 58.03µs, path_prettifier: 39.842µs, identifier_marker: 25.807µs, exclude-rules: 19.106µs, skip_dirs: 15.091µs, skip_files: 8.557µs, uniq_by_line: 2.505µs, max_from_linter: 1.253µs, cgo: 1.139µs, path_shortener: 863ns, max_same_issues: 780ns, filename_unadjuster: 642ns, max_per_file_from_linter: 559ns, severity-rules: 221ns, exclude: 187ns, diff: 165ns, sort_results: 133ns, path_prefixer: 87ns 
INFO [runner] linters took 135.107373ms with stages: gocritic: 131.834723ms 
pkg/proxy/iptables/proxier.go:1066:4: ruleguard: prefer utilnet.ParseIPSloppy() (gocritic)
			net.ParseIP("foobar") // force golangci-lint to fail
			^
INFO File cache stats: 1 entries of total size 67.5KiB 
INFO Memory: 8 samples, avg is 54.2MB, max is 90.9MB 
INFO Execution took 694.030625ms                  
```
Should fail but doesn't?

```
$ rm ruleguard_rules.go 

$ ./_output/local/bin/golangci-lint run -v ./pkg/proxy/...
INFO [config_reader] Config search paths: [./ /home/thockin/src/go/src/k8s.io/kubernetes/pkg/proxy /home/thockin/src/go/src/k8s.io/kubernetes/pkg /home/thockin/src/go/src/k8s.io/kubernetes /home/thockin/src/go/src/k8s.io /home/thockin/src/go/src /home/thockin/src/go /home/thockin/src /home/thockin /home /] 
INFO [config_reader] Used config file .golangci.yaml 
INFO [lintersdb] Active 1 linters: [gocritic]     
INFO [loader] Go packages loading at mode 575 (compiled_files|deps|exports_file|files|name|imports|types_sizes) took 547.450862ms 
INFO [runner/filename_unadjuster] Pre-built 0 adjustments in 15.642641ms 
INFO [linters context/goanalysis] analyzers took 0s with no stages 
INFO [runner] Processors filtering stat (out/in): diff: 1/1, max_from_linter: 1/1, cgo: 1/1, filename_unadjuster: 1/1, path_shortener: 1/1, severity-rules: 1/1, nolint: 1/1, uniq_by_line: 1/1, identifier_marker: 1/1, exclude: 1/1, max_per_file_from_linter: 1/1, skip_files: 1/1, autogenerated_exclude: 1/1, exclude-rules: 1/1, max_same_issues: 1/1, source_code: 1/1, path_prefixer: 1/1, sort_results: 1/1, path_prettifier: 1/1, skip_dirs: 1/1 
INFO [runner] processing took 3.513511ms with stages: nolint: 3.23721ms, source_code: 85.83µs, autogenerated_exclude: 62.202µs, path_prettifier: 41.877µs, identifier_marker: 28.896µs, exclude-rules: 19.336µs, skip_dirs: 15.184µs, skip_files: 14.751µs, uniq_by_line: 1.727µs, cgo: 1.37µs, max_from_linter: 1.366µs, path_shortener: 810ns, max_same_issues: 728ns, filename_unadjuster: 698ns, max_per_file_from_linter: 491ns, diff: 239ns, severity-rules: 216ns, exclude: 211ns, sort_results: 197ns, path_prefixer: 172ns 
INFO [runner] linters took 137.612499ms with stages: gocritic: 134.027635ms 
pkg/proxy/iptables/proxier.go:1066:4: ruleguard: prefer utilnet.ParseIPSloppy() (gocritic)
			net.ParseIP("foobar") // force golangci-lint to fail
			^
INFO File cache stats: 1 entries of total size 67.5KiB 
INFO Memory: 9 samples, avg is 56.9MB, max is 82.8MB 
INFO Execution took 705.259059ms                  
```

Should print a different error, but doesn't:

```
$ git co -- ruleguard_rules.go

$ sed -i 's/prefer/PREFER/' ruleguard_rules.go 

$ cat ruleguard_rules.go 
//+build never

package gorules

import "github.com/quasilyte/go-ruleguard/dsl"

func netParseIP(m dsl.Matcher) {
	m.Match(`net.ParseIP($_)`).Report("PREFER utilnet.ParseIPSloppy()")
}

$ ./_output/local/bin/golangci-lint run -v ./pkg/proxy/...
INFO [config_reader] Config search paths: [./ /home/thockin/src/go/src/k8s.io/kubernetes/pkg/proxy /home/thockin/src/go/src/k8s.io/kubernetes/pkg /home/thockin/src/go/src/k8s.io/kubernetes /home/thockin/src/go/src/k8s.io /home/thockin/src/go/src /home/thockin/src/go /home/thockin/src /home/thockin /home /] 
INFO [config_reader] Used config file .golangci.yaml 
INFO [lintersdb] Active 1 linters: [gocritic]     
INFO [loader] Go packages loading at mode 575 (exports_file|imports|types_sizes|deps|files|name|compiled_files) took 527.510224ms 
INFO [runner/filename_unadjuster] Pre-built 0 adjustments in 17.598719ms 
INFO [linters context/goanalysis] analyzers took 0s with no stages 
INFO [runner] Processors filtering stat (out/in): skip_dirs: 1/1, autogenerated_exclude: 1/1, nolint: 1/1, uniq_by_line: 1/1, cgo: 1/1, filename_unadjuster: 1/1, skip_files: 1/1, max_per_file_from_linter: 1/1, path_prettifier: 1/1, exclude-rules: 1/1, diff: 1/1, path_shortener: 1/1, path_prefixer: 1/1, sort_results: 1/1, identifier_marker: 1/1, exclude: 1/1, max_same_issues: 1/1, max_from_linter: 1/1, source_code: 1/1, severity-rules: 1/1 
INFO [runner] processing took 3.109847ms with stages: nolint: 2.850793ms, source_code: 78.941µs, autogenerated_exclude: 60.194µs, path_prettifier: 39.487µs, identifier_marker: 26.46µs, exclude-rules: 20.101µs, skip_dirs: 15.566µs, skip_files: 9.506µs, max_from_linter: 2.008µs, uniq_by_line: 1.717µs, cgo: 1.231µs, path_shortener: 964ns, max_same_issues: 655ns, filename_unadjuster: 630ns, exclude: 511ns, max_per_file_from_linter: 426ns, diff: 244ns, severity-rules: 208ns, sort_results: 119ns, path_prefixer: 86ns 
INFO [runner] linters took 131.961611ms with stages: gocritic: 128.786889ms 
pkg/proxy/iptables/proxier.go:1066:4: ruleguard: prefer utilnet.ParseIPSloppy() (gocritic)
			net.ParseIP("foobar") // force golangci-lint to fail
			^
INFO File cache stats: 1 entries of total size 67.5KiB 
INFO Memory: 8 samples, avg is 55.7MB, max is 86.6MB 
INFO Execution took 681.955996ms 
```

**Comment by dottedmag** [comment]
Bugs don't get fixed by themselves!

**Comment by misha-ridge** [comment]
The same issue applies to custom linter checks.

**Comment by misha-ridge** [comment]
Repository with reproducer has been moved here: https://github.com/misha-ridge/golangci-cache-ruleguard
